Win32 code refresh & audio buffer enhancements [Rebased into new PR #267]

sukibaby opened the pull request:
## There is a lot going on in this PR so please review and ask questions, I'll be happy to answer. This PR is mainly focused on Windows improvements, but also contains sound buffer improvements.

Thanks to [x64dbg](https://github.com/x64dbg/x64dbg) for helping make these discoveries possible.

### Overview
The goal of this PR is to solve the problem of unpredictable glitches on Windows hosts. On Windows, many time critical threads, including all rendering threads, are run at normal priority, meaning the system views them as very unimportant. 

When going through the code to resolve these issues, a large amount of Windows 9x compatibility code was found, and replaced with modernized code compatible with Windows 7 and up.

Bad code in the sound mixing buffer would lead to runaway memory fragmentation, causing a significant amount of RAM to be wasted. The sound mixing buffer now pre-allocates a safe amount of memory which results in a significant decrease in RAM consumption (from ~1.4GB to ~450MB on my Windows 11 PC).

There is a separate sound buffer for the Windows audio driver, and this now pre-allocates a larger amount of memory to avoid frequent reallocations.

These changes are largely interdependent so they are all in one PR.

### Testing

Since it's known that streaming with OBS or Discord will introduce stutter, I streamed into Discord while playing mod charts for half an hour, and did not observe any stutter in that time, which is not possible with the old code.

### Change overview
**GameLoop.cpp**
- Check if the OS is Windows, and if so ensure the rendering thread is set to the highest priority

**RageSound.cpp**
- Undefine `ERROR` to prevent a name mismatch when including `windows.h` elsewhere

**RageSoundMixBuffer.cpp / RageSoundMixBuffer.h**

- Set a buffer of 2MB instead of setting the buffer to nullptr and depend on it being grown by assets being loaded at the game launch
- Update some C style casts to C++ style casts
- Improve `Extend` feature to support handling a larger number of samples, and error handling in case of a memory allocation failure
    - I don't really like using std::exit in case the new buffer is nullptr, but I haven't decided what I would like to do here if it is nullptr

**RageSoundReader_Chain.cpp**

- Prevent code running needlessly on a null pointer

**RageSoundReader_ThreadedBuffer.cpp**

- BufferingThread(): Calculate unchanging values outside of the loop for better performance

**RageSoundReader_ThreadedBuffer.h**

- Include `windows.h` if the OS is Windows so that the sound buffer can be set to high priority

**arch/ArchHooks/ArchHooks_Win32.cpp**

- Set the main thread back to `ABOVE_NORMAL_PRIORITY_CLASS` to ensure the sound and rendering threads stay higher priority (reverts #228)

**arch/Sound/DSoundHelpers.cpp**

- Set the DirectSound buffer to 512KB instead of 64KB

**arch/Sound/RageSoundDriver_DSound_Software.cpp**

- Set the DirectSound decoding thread to `THREAD_PRIORITY_HIGHEST`

**arch/Sound/RageSoundDriver_WDMKS.cpp**

- Allow the buffer size to adapt to the number of channels and maximum bytes per sample
- Set one temporary buffer big enough to reuse multiple times instead of setting the buffer to ~2-4KB and reallocating it as needed
- Keep track of the starting position of `pBuf`
- Set the WDM decoding thread to `THREAD_PRIORITY_HIGHEST`

**arch/Threads/Threads_Win32.cpp**

- Rewrite `SetThreadName` to be GCC/GDB-friendly
- Update `StartThread` to use modern Windows features
- Update `MutexImpl_Win32::Lock` to use modern Windows features
- Replace all Win 9x compatibility code in `PortableSignalObjectAndWait`

geefr committed:
Add WITH_MINIMAID option to allow building on ARM

florczakraf committed:
Guard against signedness of chars and wchars on different targets

florczakraf committed:
Treat type-limits as errors in GNU builds

Pete-Lawrence committed:
Added missing lifts to stock dance Noteskins

sukibaby committed:
Win32 code refresh & audio buffer enhancements

GameLoop.cpp

Check if the OS is Windows, and if so ensure the rendering thread is set to the highest priority
RageSound.cpp

Undefine ERROR to prevent a name mismatch when including windows.h elsewhere
RageSoundMixBuffer.cpp / RageSoundMixBuffer.h

Set a buffer of 2MB instead of setting the buffer to nullptr and depend on it being grown by assets being loaded at the game launch
Update some C style casts to C++ style casts
Improve Extend feature to support handling a larger number of samples, and error handling in case of a memory allocation failure
I don't really like using std::exit in case the new buffer is nullptr, but I haven't decided what I would like to do here if it is nullptr
RageSoundReader_Chain.cpp

Prevent code running needlessly on a null pointer
RageSoundReader_ThreadedBuffer.cpp

BufferingThread(): Calculate unchanging values outside of the loop for better performance
RageSoundReader_ThreadedBuffer.h

Include windows.h if the OS is Windows so that the sound buffer can be set to high priority
arch/ArchHooks/ArchHooks_Win32.cpp

Set the main thread back to ABOVE_NORMAL_PRIORITY_CLASS to ensure the sound and rendering threads stay higher priority (reverts Change to HIGH_PRIORITY_CLASS on Windows #228)
arch/Sound/DSoundHelpers.cpp

Set the DirectSound buffer to 512KB instead of 64KB
arch/Sound/RageSoundDriver_DSound_Software.cpp

Set the DirectSound decoding thread to THREAD_PRIORITY_HIGHEST
arch/Sound/RageSoundDriver_WDMKS.cpp

Allow the buffer size to adapt to the number of channels and maximum bytes per sample
Set one temporary buffer big enough to reuse multiple times instead of setting the buffer to ~2-4KB and reallocating it as needed
Keep track of the starting position of pBuf
Set the WDM decoding thread to THREAD_PRIORITY_HIGHEST
arch/Threads/Threads_Win32.cpp

Rewrite SetThreadName to be GCC/GDB-friendly
Update StartThread to use modern Windows features
Update MutexImpl_Win32::Lock to use modern Windows features
Replace all Win 9x compatibility code in PortableSignalObjectAndWait

sukibaby committed:
Remove AltiVec code

The content of src/archutils/Darwin/VectorHelper.cpp is very old, these are AltiVec and SSE2 instructions. This is only actually used in RageSoundMixBuffer.cpp. This level of micro-optimization is not needed since platforms this old are not supported by ITGm.

sukibaby committed:
Change a lrint

sukibaby committed:
Replace src/archutils/Win32/ddk/hidpi.h

From SDK 10.0.16299

glitchybear commented:
Did the DSound and WDMKS changes have an effect even if you're using WaveOut as the audio driver for Windows? I assumed those weren't used at all since WaveOut is the default audio driver for Windows unless it fails to load for some reason (see arch/sound/RageSoundDriver.cpp and arch/arch_default.h). 

Not arguing against the changes, mostly just curious as to whether non-WaveOut drivers are actually getting used in normal Windows setups or if they're being used in a non-obvious way elsewhere.

sukibaby commented:
> Did the DSound and WDMKS changes have an effect even if you're using WaveOut as the audio driver for Windows? I assumed those weren't used at all since WaveOut is the default audio driver for Windows unless it fails to load for some reason (see arch/sound/RageSoundDriver.cpp and arch/arch_default.h).

Setting the proper thread priority for the default (WaveOut) audio handling is covered by `RageSoundReader_ThreadedBuffer`. That file is where the buffer is created ☺️ 

Of course, this also ensures that if DirectSound or WDM is used instead of (or along with) WaveOut, it'll be high priority as well, since the user doesn't choose which is used.

There are usually 3 or 4 audio related threads going on at a time, and `RageSoundReader_ThreadedBuffer` is only responsible for one of those. With that said, at this time I haven't looked into where the others are yet or if other drivers are used in those threads.

> Not arguing against the changes, mostly just curious as to whether non-WaveOut drivers are actually getting used in normal Windows setups or if they're being used in a non-obvious way elsewhere.

No worries 🙏

sukibaby committed:
Remove NTPAD check

NTpad is a driver for Windows 2000/XP which enables the use of a controller via the parallel port. This is not needed anymore and is just causing lag by constantly scanning the list of input devices on Windows.

sukibaby committed:
Actor optimizations

sukibaby commented on src/RageSoundMixBuffer.cpp:
```diff
@@ -14,14 +14,27 @@ static bool g_bVector = Vector::CheckForVector();
+	// Check if memory allocation was successful
+	if (m_pMixbuf == nullptr)
+	{
+		std::exit(EXIT_FAILURE);
```
Haven't decided what the best course of action here is, not convinced `std::exit(EXIT_FAILURE)` is ideal

teejusb replied:
Maybe ASSERT_M just indicating the you weren't able to allocate enough memory. Similar behavior but at least we'd get a log? I think it seems like a rare issue to run into considering we only need 2MB.

sukibaby committed:
Improve calc_mean function

A problem with using std::accumulate to calculate the mean is that small numbers get rounded down to zero when dealing with floating point numbers. This is solved by implementing the Kahan summation algorithm (https://en.wikipedia.org/wiki/Kahan_summation_algorithm).

teejusb commented on src/RageSound.cpp:
```diff
@@ -288,6 +288,10 @@ int RageSound::GetDataToPlay( float *pBuffer, int iFrames, std::int64_t &iStream
 		/* Read data from our source. */
 		int iGotFrames = m_pSource->RetriedRead( pBuffer + (iFramesStored * m_pSource->GetNumChannels()), iFrames, &iSourceFrame, &fRate );
 
+#ifdef ERROR
```
Can we just #undef ERROR when we include windows.h instead of everywhere?

sukibaby replied:
Sure.

sukibaby replied:
(if i'm understanding correctly) `windows.h` isn't used in RageSound, it's pulled in via GameLoop. The reason the undef is located here is to prevent a compiler error about ERROR already being defined. So if I undef ERROR in GameLoop it still needs to be undef'd here.

However.... if we look up the definition of `RageSoundReader::ERROR`, it equals `-1`, so any reason not to just change line 291 to look for `-1`? That seems to be how the rest of the code checks for return values, anyway.

https://github.com/itgmania/itgmania/blob/deb0efb8e67e8563993150eee33cd31b6f8ba058/src/RageSound.cpp#L291

teejusb commented on src/GameLoop.cpp:
```diff
@@ -21,6 +21,10 @@
 #include "RageInput.h"
 
+#ifdef _WIN32
+#include <windows.h>
```
Are we including this just for `SetThreadPriority`? Can we include `processthreadsapi.h` instead of pulling in all of windows.h which seemingly pulls in a LOT of headers.

sukibaby replied:
So,  I can make everything build correctly if I use this in GameLoop.cpp and define a separate function to handle thread boosting, and call that function from RenderThread()

```
#ifdef _WIN32
#define WIN32_LEAN_AND_MEAN
#include <processthreadsapi.h>
#endif
```

like so:
```
void BoostThreadPriorityForWin32(HANDLE hThread)
{
	SetThreadPriority(hThread, THREAD_PRIORITY_HIGHEST);
}

void ConcurrentRenderer::RenderThread()
{
#ifdef _WIN32
	HANDLE hThread = GetCurrentThread();
	BoostThreadPriorityForWin32(hThread);
#endif
```

... but then I need to include <windows.h> in `GameLoop.h`, or else this error will appear:

```
Severity	Code	Description	Project	File	Line	Suppression State	Details
Error	C1189	#error:  "No Target Architecture"	ITGmania	C:\Program Files (x86)\Windows Kits\10\Include\10.0.22621.0\um\winnt.h	169		
```

I don't love the idea of hard-coding the architecture. What would your preference be here?

sukibaby replied:
Yes. I have changed this to include `WIN32_LEAN_AND_MEAN` to reduce the number of headers it pulls in. The problem with including `processthreadsapi.h` is that `windows.h` must be included elsewhere. It's possible to include `windows.h` in `GameLoop.h` and `processthreadsapi.h` in `GameLoop.cpp`, it seems, so it seems to be like using `windows.h` with `WIN32_LEAN_AND_MEAN` is probably best?

teejusb commented on src/RageSoundMixBuffer.cpp:
```diff
@@ -14,14 +14,27 @@ static bool g_bVector = Vector::CheckForVector();
+		std::exit(EXIT_FAILURE);
+	}
+
+	memset(m_pMixbuf, 0, m_iBufSize * sizeof(float));
```
I think we can convert to std::memset here since we're changing the others anyways.

teejusb commented on src/RageSoundMixBuffer.cpp:
```diff
@@ -31,13 +44,21 @@ void RageSoundMixBuffer::SetWriteOffset( int iOffset )
+		float* m_pMixbufNew = static_cast<float*>(realloc(m_pMixbuf, sizeof(float) * realsize));
+		if (m_pMixbufNew == nullptr)
+		{
+			std::exit(EXIT_FAILURE);
```
probably ASSERT_M with a message?

teejusb commented on src/RageSoundReader_ThreadedBuffer.cpp:
```diff
@@ -220,26 +228,9 @@ void RageSoundReader_ThreadedBuffer::BufferingThread()
-		if( fTimeToSleep == 0 )
-			fTimeToSleep = float(g_iReadBlockSizeFrames) / m_iSampleRate;
-
-		if( m_Event.WaitTimeoutSupported() )
```
Are we not supporting `WaitTimeoutSupported()` anymore?

sukibaby replied:
Sorry, I've put it back.

sukibaby committed:
Update FFmpeg URL in .gitmodules

sukibaby committed:
Restore comment in Actor.cpp

sukibaby commented:
Looks like I have to toss in a few more ifdef's so non-Windows platforms will compile.

sukibaby commented:
Hmm, yesterday this branch was working on my PC but today it doesn't want to start up (crashes & exits before the songs start loading). Looking into what could have happened.

sukibaby commented:
Found the problem about why it was compiling but not launching. I was missing an if statement around an `ASSERT_M`. This is good to go now.

sukibaby committed:
add more ifdefs to fix building on non windows

sukibaby commented:
whoops, had to merge in the NTPAD fix since this PR pre-dates that PR

sukibaby committed:
Merge branch 'beta' into patch-2

sukibaby commented:
I'm closing this because the history is a giant mess. Please refer to https://github.com/itgmania/itgmania/pull/267, an identical PR.
